exchanges/stream/websocket.go

```diff
@@ -925,18 +930,66 @@ func (w *Websocket) SubscribeToChannels(channels []ChannelSubscription) error {
 	return nil
 }
 
+// AddSubscription adds a subscription to the subscription lists
+// Unlike AddSubscriptions this method will error if the subscription already exists
+func (w *Websocket) AddSubscription(c *ChannelSubscription) error {
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+	if w.subscriptions == nil {
+		w.subscriptions = subscriptionMap{}
+	}
+	key := c.EnsureKeyed()
+	if _, ok := w.subscriptions[key]; ok {
+		return ErrSubscribedAlready
+	}
+
+	n := *c // Fresh copy; we don't want to use the pointer we were given and allow encapsulation/locks to be bypassed
+	w.subscriptions[key] = &n
+
+	return nil
+}
+
+// SetSubscriptionState sets an existing subscription state
+// returns an error if the subscription is not found, or the new state is already set
+func (w *Websocket) SetSubscriptionState(c *ChannelSubscription, state ChannelState) error {
+	w.subscriptionMutex.Lock()
+	defer w.subscriptionMutex.Unlock()
+	if w.subscriptions == nil {
+		w.subscriptions = subscriptionMap{}
+	}
+	key := c.EnsureKeyed()
+	p, ok := w.subscriptions[key]
+	if !ok {
+		return ErrSubscriptionNotFound
+	}
+	if state == p.State {
+		return ErrChannelInStateAlready
+	}
+	p.State = state
```

Comment:
Agreed on validating.
Will switch to uint8 as well.
I'll just use ChannelUnsubscribing for now instead of a new array, and add a comment, because if anyone adds another it'll be self-evident from the error.
Fixed 7a122ef6